File: medplum-java/medplum-java-model/src/main/java/com/fhir/PersonGender.java
package com.fhir;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Locale;
import javax.annotation.Nonnull;

public final class PersonGender {
  public static final PersonGender OTHER = new PersonGender(Value.OTHER, "OTHER");

  public static final PersonGender FEMALE = new PersonGender(Value.FEMALE, "FEMALE");

  public static final PersonGender MALE = new PersonGender(Value.MALE, "MALE");

  private final Value value;

  private final String string;

  PersonGender(Value value, String string) {
    this.value = value;
    this.string = string;
  }

  public Value getEnumValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return this.string;
  }

  @Override
  public boolean equals(Object other) {
    return (this == other) 
      || (other instanceof PersonGender && this.string.equals(((PersonGender) other).string));
  }

  @Override
  public int hashCode() {
    return this.string.hashCode();
  }

  public <T> T accept(Visitor<T> visitor) {
    switch (value) {
      case OTHER:
        return visitor.visitOTHER();
      case FEMALE:
        return visitor.visitFEMALE();
      case MALE:
        return visitor.visitMALE();
      case UNKNOWN:
      default:
        return visitor.visitUnknown(string);
    }
  }

  @JsonCreator(
      mode = JsonCreator.Mode.DELEGATING
  )
  public static PersonGender valueOf(@Nonnull String value) {
    String upperCasedValue = value.toUpperCase(Locale.ROOT);
    switch (upperCasedValue) {
      case "OTHER":
        return OTHER;
      case "FEMALE":
        return FEMALE;
      case "MALE":
        return MALE;
      default:
        return new PersonGender(Value.UNKNOWN, upperCasedValue);
    }
  }

  public enum Value {
    MALE,

    FEMALE,

    OTHER,

    UNKNOWN
  }

  public interface Visitor<T> {
    T visitMALE();

    T visitFEMALE();

    T visitOTHER();

    T visitUnknown(String unknownType);
  }
}

File: medplum-typescript/src/model/fhir/PersonGender.ts
export type PersonGender =
    | PersonGender.Male
    | PersonGender.Female
    | PersonGender.Other;

export const PersonGender = {
    Male: "male" as PersonGender.Male,
    Female: "female" as PersonGender.Female,
    Other: "other" as PersonGender.Other,

    _visit: <Result>(
        value: PersonGender,
        visitor: PersonGender._Visitor<Result>
    ): Result => {
        switch (value) {
            case PersonGender.Male:
                return visitor.male();
            case PersonGender.Female:
                return visitor.female();
            case PersonGender.Other:
                return visitor.other();
            default:
                return visitor._unknown();
        }
    },
};

export declare namespace PersonGender {
    export type Male = "male" & {
        __Male: void;
    };
    export type Female = "female" & {
        __Female: void;
    };
    export type Other = "other" & {
        __Other: void;
    };

    export interface _Visitor<Result> {
        male: () => Result;
        female: () => Result;
        other: () => Result;
        _unknown: () => Result;
    }
}
